crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,567 @@ impl State {
+            let withdrawal_shares_redeemed = self
+                .calculate_distribute_excess_idle_withdrawal_shares_redeemed(
+                    current_block_timestamp,
+                    max_share_reserves_delta,
+                    active_lp_total_supply,
+                    withdrawal_shares_total_supply,
+                )?;
+
+            // Step 3: If none of the withdrawal shares could be redeemed, then
+            // we're done and we pay out nothing.
+            if withdrawal_shares_redeemed == fixed!(0) {
+                return Err(eyre!(
+                    "Withdrawal shares redeemed from excess idle is zero."
+                ));
+            }
+            // Step 3: Otherwise if this amount is less than or equal to the amount
+            // of withdrawal shares outstanding, then we're done and we pay out the
+            // full maximum share reserves delta.
+            else if withdrawal_shares_redeemed <= withdrawal_shares_total_supply {
+                return Ok((withdrawal_shares_redeemed, max_share_reserves_delta));
+            }
+            // Step 3: Otherwise, all of the withdrawal shares are redeemed, and we
+            // need to calculate the amount of shares the withdrawal pool should
+            // receive.
+            else {
+                withdrawal_shares_total_supply
+            }
+        };
+
+        // Step 4: Solve for the share proceeds that hold the LP share price
+        // invariant after all of the withdrawal shares are redeemed. If the
+        // calculation returns a share proceeds of zero, we can't pay out
+        // anything.
+        let share_proceeds = self.calculate_distribute_excess_idle_share_proceeds(
+            current_block_timestamp,
+            active_lp_total_supply,
+            withdrawal_shares_total_supply,
+            max_share_reserves_delta,
+            max_iterations,
+        )?;
+        if share_proceeds == fixed!(0) {
+            return Err(eyre!("Share proceeds from excess idle are zero."));
+        }
+
+        // Step 4: If the share proceeds are greater than or equal to the
+        // maximum share reserves delta that was previously calculated, then
+        // we can't distribute excess idle since we ruled out the possibility
+        // of paying out the full maximum share reserves delta in step 3.
+        if share_proceeds >= max_share_reserves_delta {
+            return Err(eyre!(
+                "Share proceeds greater than or equal to the maximum share reserves delta."
+            ));
+        }
+
+        Ok((withdrawal_shares_redeemed, share_proceeds))
+    }
+
+    /// Calculates the share proceeds to distribute to the withdrawal pool
+    /// assuming that all of the outstanding withdrawal shares will be
+    /// redeemed. The share proceeds are calculated such that the LP share
+    /// price is conserved. When we need to round, we round down to err on
+    /// the side of slightly too few shares being paid out.
```

Comment:
I think the math is helpful. Let's add this back into the comment:

```
    ///      PV(0) / l = PV(dx) / (l - dl)
    ///                =>
    ///      dl = l - l * (PV(dx) / PV(0))
```

My default is assumption is that the Rust and Solidity should match

Reply:
Nvm, I'm looking at the wrong function. IMO, this would be clearer/easier to review if we maintain the same order of functions as we had in Solidity. This should be `calculate_distribute_excess_idle_withdrawal_shares_redeemed`

Reply:
I reorganzied the functions to match the order of LPMath.sol